homeassistant/components/recorder/models.py

```diff
@@ -131,9 +132,7 @@ class States(Base):  # type: ignore
     last_changed = Column(DATETIME_TYPE, default=dt_util.utcnow)
     last_updated = Column(DATETIME_TYPE, default=dt_util.utcnow, index=True)
     created = Column(DATETIME_TYPE, default=dt_util.utcnow)
-    old_state_id = Column(
-        Integer, ForeignKey("states.state_id", ondelete="NO ACTION"), index=True
```

Comment:
Can you explain why are you removing `ondelete`?

Reply:
For Oracle DB it is invalid statement and for all HA supported DB engines "NO ACTION" (or RESTRICT) is default action and is applied even if not specified explicitly, thus it does not change anything in terms of functionality and behaviour.

Reply:
This should be fine as long as our migration logic doesn't need it.  The created table will not have the same definition, but it should be functionally the same.

Reply:
This might actually break `_update_states_table_with_foreign_key_options` in `migration.py`. We need to create a mysql table with `ondelete CASCADE` and then put it though migration to schema version 11 to make sure it gets set back to `NO ACTION`

```
    elif new_version == 11:
        _create_index(connection, "states", "ix_states_old_state_id")
        _update_states_table_with_foreign_key_options(connection, engine)
```

Ideally we add a test for this.

Reply:
Older versions used CASCADE and we migrated back to NO ACTION.  We need to test that the migration still reverts a CASCADE to NO ACTION on step 11.

Reply:
Please don't get me wrong, I am trying to figure out what to test, but I dug a bit on models.py and I cannot find any revision with ondelete CASCADE.

Previously ondelete was [SET NULL](https://github.com/home-assistant/core/blame/46e593485e516b7b48d847c061040b6503925b80/homeassistant/components/recorder/models.py#L120), before that ondelete [wasn't specified](https://github.com/home-assistant/core/blame/cb96bd9d0b46899cac433fbd32002d9ce4f95739/homeassistant/components/recorder/models.py#L109) and before that [there wasn't ForeignKey specified](https://github.com/home-assistant/core/blame/aada6a1d88a0bf93024316ae4c23350b2654a8ae/homeassistant/components/recorder/models.py#L108) so now I'm lost.

Reply:
```
Welcome to the MariaDB monitor.  Commands end with ; or \g.
Your MariaDB connection id is 2438
Server version: 10.3.21-MariaDB-log Source distribution

Copyright (c) 2000, 2018, Oracle, MariaDB Corporation Ab and others.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

MariaDB [homeassistant]> select * from schema_changes;
+-----------+----------------+---------------------+
| change_id | schema_version | changed             |
+-----------+----------------+---------------------+
|         1 |             11 | 2021-02-17 04:41:14 |
|         4 |             12 | 2021-03-11 18:27:32 |
|         5 |             13 | 2021-04-07 17:09:19 |
|         6 |             14 | 2021-04-08 02:22:53 |
+-----------+----------------+---------------------+
4 rows in set (0.032 sec)

MariaDB [homeassistant]> delete from schema_changes where change_id > 1;
Query OK, 3 rows affected (0.049 sec)

MariaDB [homeassistant]> update schema_changes set schema_version = 10;
Query OK, 1 row affected (0.031 sec)
Rows matched: 1  Changed: 1  Warnings: 0

MariaDB [homeassistant]> SHOW CREATE TABLE states;
+--------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table  | Create Table                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |
+--------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| states | CREATE TABLE `states` (
  `state_id` int(11) NOT NULL AUTO_INCREMENT,
  `domain` varchar(64) COLLATE utf8mb4_unicode_ci DEFAULT NULL,
  `entity_id` varchar(255) COLLATE utf8mb4_unicode_ci DEFAULT NULL,
  `state` varchar(255) COLLATE utf8mb4_unicode_ci DEFAULT NULL,
  `attributes` longtext COLLATE utf8mb4_unicode_ci DEFAULT NULL,
  `event_id` int(11) DEFAULT NULL,
  `last_changed` datetime(6) DEFAULT NULL,
  `last_updated` datetime(6) DEFAULT NULL,
  `created` datetime(6) DEFAULT NULL,
  `old_state_id` int(11) DEFAULT NULL,
  PRIMARY KEY (`state_id`),
  KEY `ix_states_entity_id_last_updated` (`entity_id`,`last_updated`),
  KEY `ix_states_old_state_id` (`old_state_id`),
  KEY `ix_states_event_id` (`event_id`),
  KEY `ix_states_last_updated` (`last_updated`),
  CONSTRAINT `states_ibfk_1` FOREIGN KEY (`event_id`) REFERENCES `events` (`event_id`) ON DELETE CASCADE,
  CONSTRAINT `states_ibfk_2` FOREIGN KEY (`old_state_id`) REFERENCES `states` (`state_id`) ON DELETE NO ACTION
) ENGINE=InnoDB AUTO_INCREMENT=2217889 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci |
+--------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.000 sec)


MariaDB [homeassistant]> ALTER TABLE `states` DROP FOREIGN KEY `states_ibfk_2`;
Query OK, 0 rows affected (0.019 sec)
Records: 0  Duplicates: 0  Warnings: 0

MariaDB [homeassistant]> ALTER TABLE states ADD CONSTRAINT states_ibfk_2 FOREIGN KEY (old_state_id) REFERENCES states(state_id) ON DELETE CASCADE;
Query OK, 81783 rows affected (19.229 sec)             
Records: 81783  Duplicates: 0  Warnings: 0

                                                                                                                                                                                                                                                                                                                            |
+--------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| states | CREATE TABLE `states` (
  `state_id` int(11) NOT NULL AUTO_INCREMENT,
  `domain` varchar(64) COLLATE utf8mb4_unicode_ci DEFAULT NULL,
  `entity_id` varchar(255) COLLATE utf8mb4_unicode_ci DEFAULT NULL,
  `state` varchar(255) COLLATE utf8mb4_unicode_ci DEFAULT NULL,
  `attributes` longtext COLLATE utf8mb4_unicode_ci DEFAULT NULL,
  `event_id` int(11) DEFAULT NULL,
  `last_changed` datetime(6) DEFAULT NULL,
  `last_updated` datetime(6) DEFAULT NULL,
  `created` datetime(6) DEFAULT NULL,
  `old_state_id` int(11) DEFAULT NULL,
  PRIMARY KEY (`state_id`),
  KEY `ix_states_entity_id_last_updated` (`entity_id`,`last_updated`),
  KEY `ix_states_old_state_id` (`old_state_id`),
  KEY `ix_states_event_id` (`event_id`),
  KEY `ix_states_last_updated` (`last_updated`),
  CONSTRAINT `fk_old_state_id` FOREIGN KEY (`old_state_id`) REFERENCES `states` (`state_id`) ON DELETE CASCADE,
  CONSTRAINT `states_ibfk_1` FOREIGN KEY (`event_id`) REFERENCES `events` (`event_id`) ON DELETE CASCADE,
  CONSTRAINT `states_ibfk_2` FOREIGN KEY (`old_state_id`) REFERENCES `states` (`state_id`) ON DELETE CASCADE
) ENGINE=InnoDB AUTO_INCREMENT=2217911 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci |
+--------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.000 sec)

```

Reply:
Looks like the logic won't remove it with the change 
```
CONSTRAINT `states_ibfk_2` FOREIGN KEY (`old_state_id`) REFERENCES `states` (`state_id`) ON DELETE CASCADE`
```

Its still there

Reply:
Looks like we actually shipped with `SET NULL` instead of `CASCADE` for that column so my query should have been

`ALTER TABLE states ADD CONSTRAINT states_ibfk_2 FOREIGN KEY (old_state_id) REFERENCES states(state_id) ON DELETE SET NULL;`

https://github.com/home-assistant/core/commit/337b8d279eabc4e2eb52e5b449ab89d773de6b21

Reply:
We need to do a database migration to remove the constraint from existing installs

Reply:
I pushed a commit to remove the constraint